File: python/src/revenant/core/pdf/builder.py
"""PDF signature field preparation.

High-level API for preparing PDFs with empty signature fields,
computing ByteRange hashes, and inserting CMS containers.

Low-level PDF object building is in objects.py.
Low-level rendering object construction is in render.py.
Position/geometry helpers are in position.py.
Post-sign verification is in verify.py.
"""

from __future__ import annotations

import hashlib
from typing import TYPE_CHECKING

from ...errors import PDFError
from ..appearance import (
    AppearanceData,
    SignatureImageData,
    build_appearance_stream,
    get_font,
    load_signature_image,
    make_date_str,
)
from .incremental import (
    assemble_incremental_update,
    find_page_obj_num,
    find_prev_startxref,
    find_root_obj_num,
    patch_byterange,
)
from .objects import (
    SigObjectNums,
    allocate_sig_objects,
    build_catalog_override,
    build_page_override,
)
from .position import SIG_HEIGHT, SIG_WIDTH, compute_sig_rect
from .render import (
    build_annot_widget,
    build_embedded_font_objects,
    build_form_xobjects,
    build_invisible_annot_widget,
    build_sig_dict,
)

if TYPE_CHECKING:
    from ..appearance.fonts import FontMetrics

# ── Helpers ────────────────────────────────────────────────────────────


def _to_bytes(raw: str | bytes) -> bytes:
    """Convert a raw PDF object (str or bytes) to bytes."""
    return raw if isinstance(raw, bytes) else raw.encode("latin-1")


def prepare_pdf_with_sig_field(
    pdf_bytes: bytes,
    page: int | str = 0,
    x: float | None = None,
    y: float | None = None,
    w: float = SIG_WIDTH,
    h: float = SIG_HEIGHT,
    position: str = "bottom-right",
    reason: str = "Signed with Revenant",
    name: str | None = None,
    image_path: str | None = None,
    fields: list[str] | None = None,
    visible: bool = True,
    font: str | None = None,
) -> tuple[bytes, int, int]:
    """
    Prepare a PDF with an empty signature field for hash-then-sign.

    Uses a TRUE incremental update: the original PDF bytes are preserved
    exactly, and new objects are appended after the original %%EOF.
    This is critical for validators (like e-keng) that verify the original
    file bytes are unchanged.

    Position can be set in two ways:
    - **Preset** (default): use ``position`` name like "bottom-right" or alias "br".
      Coordinates are computed from actual page dimensions.
    - **Manual**: pass explicit ``x`` and ``y`` to override the preset.

    Args:
        pdf_bytes: Raw PDF content.
        page: Page for the signature -- 0-based int, "first", or "last".
        x, y: Manual coordinates (PDF points, origin = bottom-left).
            If None, computed from ``position`` preset.
        w, h: Signature field size in PDF points.
        position: Preset name ("bottom-right", "br", etc.).
            Ignored when x/y are provided explicitly.
        reason: Signature reason string.
        name: Signer display name (used for PDF /Name entry).
        image_path: Optional path to a PNG/JPEG signature image.
        fields: Ordered display strings for the signature appearance.
            If None, defaults to [name, auto_date].
        visible: If False, create an invisible signature (no visual
            appearance on the page, /Rect [0 0 0 0]).
        font: Font registry key (e.g. "noto-sans", "ghea-grapalat").
            If None, uses the default font.

    Returns:
        (pdf_bytes, contents_hex_offset, contents_hex_length)
    """
    # ── Read-only analysis of the original PDF ──────────────────
    root_obj_num, root_gen = find_root_obj_num(pdf_bytes)
    page_obj_num, page_w, page_h, _has_annots, existing_annots = find_page_obj_num(
        pdf_bytes, root_obj_num, page
    )
    prev_xref, prev_size, trailer_extra = find_prev_startxref(pdf_bytes)

    if visible:
        raw_objects, new_size = _prepare_visible(
            pdf_bytes,
            prev_size,
            root_obj_num,
            page_obj_num,
            existing_annots,
            page_w,
            page_h,
            x,
            y,
            w,
            h,
            position,
            reason,
            name,
            image_path,
            fields,
            font,
        )
    else:
        raw_objects, new_size = _prepare_invisible(
            pdf_bytes,
            prev_size,
            root_obj_num,
            page_obj_num,
            existing_annots,
            reason,
            name,
        )

    # ── Assemble incremental update ────────────────────────────
    full_pdf = assemble_incremental_update(
        pdf_bytes=pdf_bytes,
        raw_objects=raw_objects,
        new_size=new_size,
        prev_xref=prev_xref,
        root_obj_num=root_obj_num,
        root_gen=root_gen,
        trailer_extra=trailer_extra,
    )

    # ── Patch the ByteRange and return offsets ─────────────────
    return patch_byterange(full_pdf, len(pdf_bytes))


def _prepare_visible(
    pdf_bytes: bytes,
    prev_size: int,
    root_obj_num: int,
    page_obj_num: int,
    existing_annots: list[str],
    page_w: float,
    page_h: float,
    x: float | None,
    y: float | None,
    w: float,
    h: float,
    position: str,
    reason: str,
    name: str | None,
    image_path: str | None,
    fields: list[str] | None,
    font_key: str | None = None,
) -> tuple[list[tuple[bytes, int]], int]:
    """Build PDF objects for a visible signature."""
    font_obj = get_font(font_key)

    # Resolve signature position
    if x is None or y is None:
        x, y, w, h = compute_sig_rect(page_w, page_h, position, w, h)

    # Load signature image first (need dimensions for aspect-ratio-correct layout)
    img_data = None
    image_aspect = None
    if image_path is not None:
        img_data = load_signature_image(image_path)
        if img_data["height"] > 0:
            image_aspect = img_data["width"] / img_data["height"]

    # Build appearance stream
    display_fields = fields
    if display_fields is None:
        display_fields = [name or "Digital Signature", f"Date: {make_date_str()}"]
    ap_info = build_appearance_stream(
        width=w,
        height=h,
        fields=display_fields,
        has_image=(image_path is not None),
        font=font_obj,
        image_aspect=image_aspect,
    )

    # Allocate new object numbers
    has_image = img_data is not None
    has_smask = has_image and img_data["smask"] is not None
    obj_nums = allocate_sig_objects(prev_size, has_image, has_smask, visible=True)

    raw_objects = _build_all_objects(
        pdf_bytes=pdf_bytes,
        obj_nums=obj_nums,
        root_obj_num=root_obj_num,
        page_obj_num=page_obj_num,
        existing_annots=existing_annots,
        x=x,
        y=y,
        w=w,
        h=h,
        reason=reason,
        name=name,
        ap_info=ap_info,
        img_data=img_data,
        font_metrics=font_obj.metrics,
    )
    return raw_objects, obj_nums["new_size"]


def _prepare_invisible(
    pdf_bytes: bytes,
    prev_size: int,
    root_obj_num: int,
    page_obj_num: int,
    existing_annots: list[str],
    reason: str,
    name: str | None,
) -> tuple[list[tuple[bytes, int]], int]:
    """Build PDF objects for an invisible signature (no visual appearance)."""
    obj_nums = allocate_sig_objects(prev_size, has_image=False, has_smask=False, visible=False)

    sig_dict_raw = build_sig_dict(obj_nums["sig"], reason, name)
    annot_raw = build_invisible_annot_widget(obj_nums, page_obj_num)

    annots_list = " ".join(existing_annots) if existing_annots else ""
    if annots_list:
        annots_list += " "
    annots_list += f"{obj_nums['annot']} 0 R"

    page_override = build_page_override(pdf_bytes, page_obj_num, annots_list)
    catalog_override = build_catalog_override(pdf_bytes, root_obj_num, obj_nums["annot"])

    raw_objects = [
        (_to_bytes(sig_dict_raw), obj_nums["sig"]),
        (_to_bytes(annot_raw), obj_nums["annot"]),
        (_to_bytes(page_override), page_obj_num),
        (_to_bytes(catalog_override), root_obj_num),
    ]
    return raw_objects, obj_nums["new_size"]


def _build_all_objects(
    pdf_bytes: bytes,
    obj_nums: SigObjectNums,
    root_obj_num: int,
    page_obj_num: int,
    existing_annots: list[str],
    x: float,
    y: float,
    w: float,
    h: float,
    reason: str,
    name: str | None,
    ap_info: AppearanceData,
    img_data: SignatureImageData | None,
    font_metrics: FontMetrics | None = None,
) -> list[tuple[bytes, int]]:
    """Build all raw PDF objects for the incremental update.

    Returns:
        List of (raw_bytes, obj_num) tuples in append order.
    """
    if font_metrics is None:
        font_metrics = get_font().metrics
    sig_dict_raw = build_sig_dict(obj_nums["sig"], reason, name)
    font_objects = build_embedded_font_objects(obj_nums, font_metrics)

    n0_raw, n2_raw, frm_raw, ap_raw, img_raw, smask_raw = build_form_xobjects(
        obj_nums, w, h, ap_info, img_data
    )
    annot_raw = build_annot_widget(obj_nums, page_obj_num, x, y, w, h)

    # Updated page object (override with /Annots added)
    annots_list = " ".join(existing_annots) if existing_annots else ""
    if annots_list:
        annots_list += " "
    annots_list += f"{obj_nums['annot']} 0 R"

    page_override = build_page_override(pdf_bytes, page_obj_num, annots_list)
    catalog_override = build_catalog_override(pdf_bytes, root_obj_num, obj_nums["annot"])

    return _collect_objects(
        sig_dict_raw=sig_dict_raw,
        annot_raw=annot_raw,
        font_objects=font_objects,
        n0_raw=n0_raw,
        n2_raw=n2_raw,
        frm_raw=frm_raw,
        ap_raw=ap_raw,
        img_raw=img_raw,
        smask_raw=smask_raw,
        page_override=page_override,
        catalog_override=catalog_override,
        obj_nums=obj_nums,
        page_obj_num=page_obj_num,
        root_obj_num=root_obj_num,
    )


def _collect_objects(
    sig_dict_raw: str,
    annot_raw: str,
    font_objects: list[tuple[bytes, int]],
    n0_raw: bytes,
    n2_raw: bytes,
    frm_raw: bytes,
    ap_raw: bytes,
    img_raw: bytes,
    smask_raw: bytes,
    page_override: str,
    catalog_override: str,
    obj_nums: SigObjectNums,
    page_obj_num: int,
    root_obj_num: int,
) -> list[tuple[bytes, int]]:
    """Collect all raw objects into (bytes, obj_num) tuples in append order."""
    n0_num = obj_nums["n0"]
    n2_num = obj_nums["n2"]
    frm_num = obj_nums["frm"]
    ap_num = obj_nums["ap"]
    if n0_num is None or n2_num is None or frm_num is None or ap_num is None:
        raise PDFError("Visible signature requires n0/n2/frm/ap object numbers, but some are None")

    result: list[tuple[bytes, int]] = []
    result.append((_to_bytes(sig_dict_raw), obj_nums["sig"]))
    result.append((_to_bytes(annot_raw), obj_nums["annot"]))
    result.extend(font_objects)
    result.append((n0_raw, n0_num))
    result.append((n2_raw, n2_num))
    result.append((frm_raw, frm_num))
    result.append((ap_raw, ap_num))
    img_num = obj_nums["img"]
    if img_raw and img_num is not None:
        result.append((img_raw, img_num))
    smask_num = obj_nums["smask"]
    if smask_raw and smask_num is not None:
        result.append((smask_raw, smask_num))
    result.append((_to_bytes(page_override), page_obj_num))
    result.append((_to_bytes(catalog_override), root_obj_num))
    return result


def compute_byterange_hash(pdf_bytes: bytes, hex_start: int, hex_len: int) -> bytes:
    """Compute SHA-1 hash of the ByteRange (everything except the Contents hex)."""
    end = hex_start + hex_len
    if hex_start <= 0 or end + 1 > len(pdf_bytes):
        raise PDFError(
            f"Invalid hex range: start={hex_start}, len={hex_len}, pdf_size={len(pdf_bytes)}"
        )
    if pdf_bytes[hex_start - 1 : hex_start] != b"<":
        raise PDFError("Malformed Contents field: expected '<' before hex data")
    if pdf_bytes[end : end + 1] != b">":
        raise PDFError("Malformed Contents field: expected '>' after hex data")

    before = pdf_bytes[:hex_start]
    after = pdf_bytes[end + 1 :]  # +1 for closing ">"
    h = hashlib.sha1()
    h.update(before)
    h.update(after)
    return h.digest()


def insert_cms(pdf_bytes: bytes, hex_start: int, hex_len: int, cms_der: bytes) -> bytes:
    """Insert the CMS DER bytes as a hex string into the reserved Contents."""
    cms_hex = cms_der.hex()
    if len(cms_hex) > hex_len:
        raise PDFError(f"CMS too large: {len(cms_hex)} hex chars > {hex_len} reserved")
    cms_hex_padded = cms_hex + "0" * (hex_len - len(cms_hex))

    result = bytearray(pdf_bytes)
    result[hex_start : hex_start + hex_len] = cms_hex_padded.encode("ascii")
    return bytes(result)

File: typescript/src/core/pdf/builder.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * PDF signature field preparation.
 *
 * High-level API for preparing PDFs with empty signature fields,
 * computing ByteRange hashes, and inserting CMS containers.
 *
 * Low-level PDF object building is in objects.ts.
 * Low-level rendering object construction is in render.ts.
 * Position/geometry helpers are in position.ts.
 * Post-sign verification is in verify.ts.
 */

import { createHash } from "node:crypto";

import { PDFError } from "../../errors.js";
import { bytesToHex } from "../../utils.js";
import type { FontMetrics } from "../appearance/fonts.js";
import {
  type AppearanceData,
  buildAppearanceStream,
  getFont,
  loadSignatureImage,
  makeDateStr,
  type SignatureImageData,
} from "../appearance/index.js";
import {
  assembleIncrementalUpdate,
  findPageObjNum,
  findPrevStartxref,
  findRootObjNum,
  patchByterange,
} from "./incremental.js";
import {
  allocateSigObjects,
  buildCatalogOverride,
  buildPageOverride,
  type SigObjectNums,
} from "./objects.js";
import { computeSigRect, SIG_HEIGHT, SIG_WIDTH } from "./position.js";
import {
  buildAnnotWidget,
  buildEmbeddedFontObjects,
  buildFormXobjects,
  buildInvisibleAnnotWidget,
  buildSigDict,
  type FormXobjects,
} from "./render.js";

// -- Helpers ------------------------------------------------------------------

function toBytes(raw: string | Uint8Array): Uint8Array {
  if (raw instanceof Uint8Array) return raw;
  return new TextEncoder().encode(raw);
}

// -- Public API ---------------------------------------------------------------

export interface PrepareOptions {
  /** Page for the signature -- 0-based int, "first", or "last". */
  page?: number | string;
  /** Manual X coordinate (PDF points, origin = bottom-left). */
  x?: number | null;
  /** Manual Y coordinate (PDF points, origin = bottom-left). */
  y?: number | null;
  /** Signature field width in PDF points. */
  width?: number;
  /** Signature field height in PDF points. */
  height?: number;
  /** Preset name ("bottom-right", "br", etc.). Ignored when x/y provided. */
  position?: string;
  /** Signature reason string. */
  reason?: string;
  /** Signer display name (used for PDF /Name entry). */
  name?: string | null;
  /** Path to a PNG/JPEG signature image. */
  imagePath?: string | null;
  /** Ordered display strings for the signature appearance. */
  fields?: string[] | null;
  /** If false, create an invisible signature (no visual appearance). */
  visible?: boolean;
  /** Font registry key (e.g. "noto-sans", "ghea-grapalat"). */
  font?: string | null;
}

export interface PrepareResult {
  pdf: Uint8Array;
  hexStart: number;
  hexLen: number;
}

/**
 * Prepare a PDF with an empty signature field for hash-then-sign.
 *
 * Uses a TRUE incremental update: the original PDF bytes are preserved
 * exactly, and new objects are appended after the original %%EOF.
 */
export async function preparePdfWithSigField(
  pdfBytes: Uint8Array,
  options: PrepareOptions = {},
): Promise<PrepareResult> {
  const {
    page = 0,
    x = null,
    y = null,
    width = SIG_WIDTH,
    height = SIG_HEIGHT,
    position = "bottom-right",
    reason = "",
    name = null,
    imagePath = null,
    fields = null,
    visible = true,
    font = null,
  } = options;

  // Read-only analysis of the original PDF
  const rootInfo = findRootObjNum(pdfBytes);
  const pageInfo = await findPageObjNum(pdfBytes, page);
  const prevInfo = await findPrevStartxref(pdfBytes);

  let rawObjects: Array<{ raw: Uint8Array; objNum: number }>;
  let newSize: number;

  if (visible) {
    const result = await prepareVisible({
      pdfBytes,
      prevSize: prevInfo.maxSize,
      rootObjNum: rootInfo.objNum,
      pageObjNum: pageInfo.pageObjNum,
      existingAnnots: pageInfo.existingAnnots,
      pageW: pageInfo.pageWidth,
      pageH: pageInfo.pageHeight,
      xPos: x,
      yPos: y,
      width,
      height,
      position,
      reason,
      name,
      imagePath,
      displayFields: fields,
      fontKey: font,
    });
    rawObjects = result.rawObjects;
    newSize = result.newSize;
  } else {
    const result = await prepareInvisible(
      pdfBytes,
      prevInfo.maxSize,
      rootInfo.objNum,
      pageInfo.pageObjNum,
      pageInfo.existingAnnots,
      reason,
      name,
    );
    rawObjects = result.rawObjects;
    newSize = result.newSize;
  }

  // Assemble incremental update
  const fullPdf = assembleIncrementalUpdate(
    pdfBytes,
    rawObjects,
    newSize,
    prevInfo.prevXref,
    rootInfo.objNum,
    rootInfo.genNum,
    prevInfo.trailerExtra,
    prevInfo.useXrefStream,
  );

  // Patch the ByteRange and return offsets
  return patchByterange(fullPdf, pdfBytes.length);
}

/**
 * Compute SHA-1 hash of the ByteRange (everything except the Contents hex).
 */
export function computeByterangeHash(
  pdfBytes: Uint8Array,
  hexStart: number,
  hexLen: number,
): Uint8Array {
  const end = hexStart + hexLen;
  if (hexStart <= 0 || end + 1 > pdfBytes.length) {
    throw new PDFError(
      `Invalid hex range: start=${hexStart}, len=${hexLen}, pdf_size=${pdfBytes.length}`,
    );
  }
  if (pdfBytes[hexStart - 1] !== 0x3c) {
    throw new PDFError("Malformed Contents field: expected '<' before hex data");
  }
  if (pdfBytes[end] !== 0x3e) {
    throw new PDFError("Malformed Contents field: expected '>' after hex data");
  }

  const before = pdfBytes.slice(0, hexStart);
  const after = pdfBytes.slice(end + 1);
  const hash = createHash("sha1");
  hash.update(before);
  hash.update(after);
  return new Uint8Array(hash.digest());
}

/**
 * Insert the CMS DER bytes as a hex string into the reserved Contents.
 */
export function insertCms(
  pdfBytes: Uint8Array,
  hexStart: number,
  hexLen: number,
  cmsDer: Uint8Array,
): Uint8Array {
  const cmsHex = bytesToHex(cmsDer);
  if (cmsHex.length > hexLen) {
    throw new PDFError(`CMS too large: ${cmsHex.length} hex chars > ${hexLen} reserved`);
  }
  const cmsHexPadded = cmsHex + "0".repeat(hexLen - cmsHex.length);

  const result = new Uint8Array(pdfBytes);
  const hexBytes = new TextEncoder().encode(cmsHexPadded);
  result.set(hexBytes, hexStart);
  return result;
}

// -- Visible signature --------------------------------------------------------

interface VisibleSigParams {
  pdfBytes: Uint8Array;
  prevSize: number;
  rootObjNum: number;
  pageObjNum: number;
  existingAnnots: string[];
  pageW: number;
  pageH: number;
  xPos: number | null;
  yPos: number | null;
  width: number;
  height: number;
  position: string;
  reason: string;
  name: string | null;
  imagePath: string | null;
  displayFields: string[] | null;
  fontKey: string | null;
}

async function prepareVisible(
  params: VisibleSigParams,
): Promise<{ rawObjects: Array<{ raw: Uint8Array; objNum: number }>; newSize: number }> {
  const fontObj = await getFont(params.fontKey);

  // Resolve signature position
  let x = params.xPos;
  let y = params.yPos;
  let w = params.width;
  let h = params.height;
  if (x === null || y === null) {
    const rect = computeSigRect(params.pageW, params.pageH, params.position, w, h);
    x = rect.x;
    y = rect.y;
    w = rect.width;
    h = rect.height;
  }

  // Load signature image (need dimensions for aspect-ratio-correct layout)
  let imgData: SignatureImageData | null = null;
  let imageAspect: number | null = null;
  if (params.imagePath !== null) {
    imgData = await loadSignatureImage(params.imagePath);
    if (imgData.height > 0) {
      imageAspect = imgData.width / imgData.height;
    }
  }

  // Build appearance stream
  const fields = params.displayFields ?? [
    params.name ?? "Digital Signature",
    `Date: ${makeDateStr()}`,
  ];
  const apInfo = await buildAppearanceStream(
    w,
    h,
    fields,
    params.imagePath !== null,
    fontObj,
    imageAspect,
  );

  // Allocate new object numbers
  const hasImage = imgData !== null;
  const hasSmask = imgData !== null && imgData.smask !== null;
  const objNums = allocateSigObjects(params.prevSize, hasImage, hasSmask, true);

  const rawObjects = await buildAllObjects(
    params.pdfBytes,
    objNums,
    params.rootObjNum,
    params.pageObjNum,
    params.existingAnnots,
    x,
    y,
    w,
    h,
    params.reason,
    params.name,
    apInfo,
    imgData,
    fontObj.metrics,
  );

  return { rawObjects, newSize: objNums.newSize };
}

// -- Invisible signature ------------------------------------------------------

async function prepareInvisible(
  pdfBytes: Uint8Array,
  prevSize: number,
  rootObjNum: number,
  pageObjNum: number,
  existingAnnots: string[],
  reason: string,
  name: string | null,
): Promise<{ rawObjects: Array<{ raw: Uint8Array; objNum: number }>; newSize: number }> {
  const objNums = allocateSigObjects(prevSize, false, false, false);

  const sigDictRaw = buildSigDict(objNums.sig, reason, name);
  const annotRaw = buildInvisibleAnnotWidget(objNums, pageObjNum);

  let annotsList = existingAnnots.length > 0 ? existingAnnots.join(" ") : "";
  if (annotsList) annotsList += " ";
  annotsList += `${objNums.annot} 0 R`;

  const pageOverride = await buildPageOverride(pdfBytes, pageObjNum, annotsList);
  const catalogOverride = await buildCatalogOverride(pdfBytes, rootObjNum, objNums.annot);

  const rawObjects: Array<{ raw: Uint8Array; objNum: number }> = [
    { raw: toBytes(sigDictRaw), objNum: objNums.sig },
    { raw: toBytes(annotRaw), objNum: objNums.annot },
    { raw: toBytes(pageOverride), objNum: pageObjNum },
    { raw: toBytes(catalogOverride), objNum: rootObjNum },
  ];

  return { rawObjects, newSize: objNums.newSize };
}

// -- Object assembly ----------------------------------------------------------

async function buildAllObjects(
  pdfBytes: Uint8Array,
  objNums: SigObjectNums,
  rootObjNum: number,
  pageObjNum: number,
  existingAnnots: string[],
  x: number,
  y: number,
  w: number,
  h: number,
  reason: string,
  name: string | null,
  apInfo: AppearanceData,
  imgData: SignatureImageData | null,
  fontMetrics: FontMetrics,
): Promise<Array<{ raw: Uint8Array; objNum: number }>> {
  const sigDictRaw = buildSigDict(objNums.sig, reason, name);
  const fontObjects = buildEmbeddedFontObjects(objNums, fontMetrics);
  const xobjects = buildFormXobjects(objNums, w, h, apInfo, imgData);
  const annotRaw = buildAnnotWidget(objNums, pageObjNum, x, y, w, h);

  let annotsList = existingAnnots.length > 0 ? existingAnnots.join(" ") : "";
  if (annotsList) annotsList += " ";
  annotsList += `${objNums.annot} 0 R`;

  const pageOverride = await buildPageOverride(pdfBytes, pageObjNum, annotsList);
  const catalogOverride = await buildCatalogOverride(pdfBytes, rootObjNum, objNums.annot);

  return collectObjects(
    sigDictRaw,
    annotRaw,
    fontObjects,
    xobjects,
    pageOverride,
    catalogOverride,
    objNums,
    pageObjNum,
    rootObjNum,
  );
}

function collectObjects(
  sigDictRaw: string,
  annotRaw: string,
  fontObjects: Array<{ raw: Uint8Array; objNum: number }>,
  xobjects: FormXobjects,
  pageOverride: string,
  catalogOverride: string,
  objNums: SigObjectNums,
  pageObjNum: number,
  rootObjNum: number,
): Array<{ raw: Uint8Array; objNum: number }> {
  if (objNums.n0 === null || objNums.n2 === null || objNums.frm === null || objNums.ap === null) {
    throw new PDFError("Visible signature requires n0/n2/frm/ap object numbers");
  }

  const result: Array<{ raw: Uint8Array; objNum: number }> = [];
  result.push({ raw: toBytes(sigDictRaw), objNum: objNums.sig });
  result.push({ raw: toBytes(annotRaw), objNum: objNums.annot });
  result.push(...fontObjects);
  result.push({ raw: xobjects.n0, objNum: objNums.n0 });
  result.push({ raw: xobjects.n2, objNum: objNums.n2 });
  result.push({ raw: xobjects.frm, objNum: objNums.frm });
  result.push({ raw: xobjects.ap, objNum: objNums.ap });

  if (xobjects.img.length > 0 && objNums.img !== null) {
    result.push({ raw: xobjects.img, objNum: objNums.img });
  }
  if (xobjects.smask.length > 0 && objNums.smask !== null) {
    result.push({ raw: xobjects.smask, objNum: objNums.smask });
  }

  result.push({ raw: toBytes(pageOverride), objNum: pageObjNum });
  result.push({ raw: toBytes(catalogOverride), objNum: rootObjNum });

  return result;
}
